File: TugasPD3.py
def hitung_bobot(nilai_huruf):
    nilai_huruf = nilai_huruf.upper()
    dict_nilai = {'A': 4.0, 'B': 3.0, 'C': 2.0, 'D': 1.0, 'E': 0.0}
    return dict_nilai.get(nilai_huruf, 0.0)

def tentukan_maks_sks(ip):
    if ip >= 3.00:
        return 24
    elif ip >= 2.50:
        return 21
    elif ip >= 2.00:
        return 18
    else:
        return 15

def main():
    print("=== Program Penghitung IP Semester ===")
    
    try:
        jumlah_mk = int(input("Masukkan jumlah mata kuliah: "))
        total_poin = 0
        total_sks = 0

        for i in range(jumlah_mk):
            print(f"\nMata Kuliah ke-{i+1}")
            sks = int(input(f"  SKS: "))
            nilai = input(f"  Nilai Huruf (A/B/C/D/E): ")
            
            bobot = hitung_bobot(nilai)
            total_poin += (bobot * sks)
            total_sks += sks

        if total_sks == 0:
            print("\nError: Total SKS tidak boleh nol.")
            return

        ip = total_poin / total_sks
        maks_sks = tentukan_maks_sks(ip)

        print("\n" + "="*30)
        print(f"Total SKS Diambil : {total_sks}")
        print(f"IP Semester Anda  : {ip:.2f}")
        print(f"Maksimal SKS yang dapat diambil selanjutnya: {maks_sks} SKS")
        print("="*30)

    except ValueError:
        print("Input tidak valid. Harap masukkan angka untuk SKS dan jumlah MK.")

if __name__ == "__main__":
    main()

File: TugasPD3.java
import java.util.Scanner;

public class TugasPD3 {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);

        System.out.println("=== PROGRAM PENGHITUNG IP SEMESTER ===");
        
        try {
            System.out.print("Masukkan jumlah mata kuliah: ");
            int jumlahMK = input.nextInt();

            double totalPoin = 0;
            int totalSKS = 0;

            for (int i = 1; i <= jumlahMK; i++) {
                System.out.println("\nMata Kuliah ke-" + i);
                System.out.print("  SKS : ");
                int sks = input.nextInt();
                
                System.out.print("  Nilai Huruf (A/B/C/D/E): ");
                String nilaiHuruf = input.next().toUpperCase();

                double bobot = tentukanBobot(nilaiHuruf);
                
                // Rumus: Sum (Ki * Ni)
                totalPoin += (sks * bobot);
                // Rumus: Sum (Ki)
                totalSKS += sks;
            }

            if (totalSKS == 0) {
                System.out.println("Error: Total SKS tidak boleh nol.");
            } else {
                // Perhitungan IP sesuai rumus di gambar
                double ip = totalPoin / totalSKS;
                int maksSKS = tentukanMaksSKS(ip);

                System.out.println("\n" + "=".repeat(35));
                System.out.println("HASIL PERHITUNGAN");
                System.out.println("=".repeat(35));
                System.out.printf("Total SKS Diambil : %d\n", totalSKS);
                System.out.printf("IP Semester Anda  : %.2f\n", ip);
                System.out.println("Batas SKS Maksimal: " + maksSKS + " SKS");
                System.out.println("=".repeat(35));
            }

        } catch (Exception e) {
            System.out.println("Terjadi kesalahan input. Pastikan memasukkan data yang benar.");
        } finally {
            input.close();
        }
    }

    // Fungsi untuk menentukan Ni (Bobot Nilai)
    public static double tentukanBobot(String nilai) {
        switch (nilai) {
            case "A" : return 4.0;
            case "B+": return 3.5;
            case "B" : return 3.0;
            case "C+": return 2.5;
            case "C" : return 2.0;
            case "D" : return 1.0;
            default : return 0.0; //dianggap E 
        }
    }

    // Fungsi untuk menentukan batas SKS semester berikutnya
    public static int tentukanMaksSKS(double ip) {
        if (ip >= 3.00) return 24;
        else if (ip >= 2.50) return 21;
        else if (ip >= 2.00) return 18;
        else return 15;
    }
}